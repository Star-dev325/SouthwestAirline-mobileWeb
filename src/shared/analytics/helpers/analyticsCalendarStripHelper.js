// @flow
import dayjs from 'dayjs';

import type { BoundType } from 'src/shared/flow-typed/flightBound.types';
import { INBOUND, OUTBOUND } from 'src/shared/constants/flightBoundTypes';

export const calculateDateDiffs = (
  newDate: string,
  previousDate: string,
  direction: BoundType,
  currentDiffs: { [BoundType]: string } = {}
): { [BoundType]: string } => {
  const diff = dayjs(newDate).diff(previousDate, 'days');
  const sign = Math.sign(diff) === 1 ? '+' : '';
  const diffs = { ...currentDiffs };

  diffs[direction] = `${sign}${diff}`;

  return diffs;
};

export const calculateOverrideDateDiffs = (
  newDate: string,
  previousDate: string,
  direction: BoundType,
  isOverrideEndDate: boolean,
  previousDiffs: { [BoundType]: string } = {}
): { [BoundType]: string } => {
  let diffs = {};

  if (direction === OUTBOUND) {
    if (isOverrideEndDate) {
      diffs = calculateDateDiffs(newDate, previousDate, INBOUND, previousDiffs);
    } else {
      diffs = {};
    }
  }

  return diffs;
};
