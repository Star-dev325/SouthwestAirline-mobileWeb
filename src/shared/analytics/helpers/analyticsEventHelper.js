import _ from 'lodash';

const pageIdentifierTransformer = [
  { matcher: /^\/$|^\/\?/, id: 'home' },
  { matcher: /\/flight-status\/(?!.*html).*/, id: 'flight-status-results' },

  { matcher: /\/flight-status\?.*_modal=originAirport/, id: 'flight-status-from' },
  { matcher: /\/flight-status\?.*_modal=destinationAirport/, id: 'flight-status-to' },

  { matcher: /\/air\/booking\/shopping\?.*_modal=from/, id: 'air-booking-shopping-from' },
  { matcher: /\/air\/booking\/shopping\?.*_modal=to/, id: 'air-booking-shopping-to' },
  { matcher: /\/air\/booking\/shopping\?.*_modal=airBookingCalendar/, id: 'air-booking-shopping-dates' },

  { matcher: /\/air\/change\/shopping\?.*_modal=from/, id: 'air-change-shopping-from' },
  { matcher: /\/air\/change\/shopping\?.*_modal=to/, id: 'air-change-shopping-to' },
  { matcher: /\/air\/change\/shopping\?.*_modal=airBookingCalendar/, id: 'air-change-shopping-dates' },

  { matcher: /\/car\/booking\?.*_modal=pickUpModal/, id: 'car-booking-pick-up' },
  { matcher: /\/car\/booking\?.*_modal=dropOffModal/, id: 'car-booking-drop-off' },
  { matcher: /\/car\/booking\?.*_modal=carBookingCalendar/, id: 'car-booking-dates' },

  { matcher: /\/check-in\/reservation\/\w{6}$/, id: 'checkin-view-reservation' }
];

const _matchPath = (path) =>
  _.get(
    _.find(pageIdentifierTransformer, (t) => path.match(t.matcher)),
    'id'
  );

export const transformPath = (path, search = null) => {
  if (!path) {
    return '';
  }
  let toMatchPath = path;

  if (search) {
    toMatchPath = `${path}${search}`;
  }

  return _matchPath(toMatchPath) || path.replace(/\//g, '-').slice(1);
};

export const raiseEvent = (eventName) => {
  window._trackAnalytics && window._trackAnalytics.event();

  const customEvent = new window.CustomEvent(eventName);

  document.body.dispatchEvent(customEvent);
};

export const raiseSatelliteEvent = (eventName, detail) => {
  if (window._satellite) {
    detail ? window._satellite.track(eventName, detail) : window._satellite.track(eventName);
  }
};

export const customerMessageAnalytics = (messages) => {
  if (messages) {
    const messageKeys = _.chain(messages).map('key').compact().join().value();

    _.set(window, 'data_a.message.customer', messageKeys);
    _.set(window, 'data_a.message.customerdisplay', '1');
  }
};

export const getPageIdentifier = (location) => {
  const { pathname = '', search = '' } = location || {};

  return transformPath(pathname, search);
};
