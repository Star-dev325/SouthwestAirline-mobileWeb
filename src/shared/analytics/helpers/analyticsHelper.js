import _ from 'lodash';

export const generateUpdatedFlowStoreForAnalytics = (flowSelectors, state, actionType) =>
  _.chain(flowSelectors)
    .pickBy((selectorObj) => _.includes(selectorObj.actions, actionType))
    .mapValues((selectorObj) => selectorObj.selector(state))
    .value();

export const generateFlowActionListForAnalytics = (flowSelectors) =>
  _.chain(flowSelectors).mapValues('actions').values().flatten().uniq().value();

export const generateUpdatedStoresForAnalytics = (generators, state, actionType) =>
  _.reduce(
    generators,
    (result, generator, key) => {
      _.each(generator(state, actionType), (subValue, subKey) => {
        result[`${key}.${subKey}`] = subValue;
      });

      return result;
    },
    {}
  );
