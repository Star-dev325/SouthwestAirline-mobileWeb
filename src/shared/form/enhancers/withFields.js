// @flow

import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import type { ComponentType } from 'react';
import { getHocDisplayName } from 'src/shared/enhancers/hocHelper';

type Props = {
  names: string[]
};

const withFields = (Component: ComponentType<*>) => {
  class WithFields extends React.Component<Props> {
    static contextTypes = {
      form: PropTypes.shape({
        onChange: PropTypes.func.isRequired,
        clearError: PropTypes.func.isRequired,
        register: PropTypes.func.isRequired,
        unregister: PropTypes.func.isRequired,
        errors: PropTypes.object.isRequired,
        formData: PropTypes.object.isRequired
      })
    };

    _onChange = (name: string, ...args: *) => {
      const { names } = this.props;
      const { onChange } = this.context.form;

      if (_.includes(names, name)) {
        onChange(name, ...args);
      }
    };

    render() {
      const { names, ...restProps } = this.props;
      const { formData } = this.context.form;

      return (
        // $FlowFixMe - Flow expects all of the wrapped components' props to be typed in this HOC, too
        <Component formData={_.pick(formData, names)} onChange={this._onChange} {...restProps} />
      );
    }
  }
  WithFields.displayName = getHocDisplayName('WithFields', Component);

  return WithFields;
};

export default withFields;
