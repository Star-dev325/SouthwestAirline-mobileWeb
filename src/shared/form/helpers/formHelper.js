// @flow
const repeatStr = (str, times = 1) => {
  let result = str;

  while (times > 1) {
    result += str;
    times--;
  }

  return result;
};

export const getMaskProps = function (mask: string | { rule: string, repeat?: number, maskChar?: string }) {
  const defaultMask = {
    repeat: 1,
    maskChar: null
  };

  const maskConfigs = Object.assign({}, defaultMask, typeof mask === 'string' ? { rule: mask } : mask);

  const finalMaskStr = repeatStr(maskConfigs.rule, maskConfigs.repeat);

  return {
    mask: finalMaskStr,
    maskChar: maskConfigs.maskChar,
    maxLength: finalMaskStr.length // mask can't work properly in some android device so use maxLength to limit the user input, see bug#2026
  };
};
