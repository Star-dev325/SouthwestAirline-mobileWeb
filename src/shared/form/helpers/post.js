import _ from 'lodash';
import BrowserObject from 'src/shared/helpers/browserObject';

const { document } = BrowserObject;

const post = (formProps, params) => {
  const createForm = () => {
    const form = document.createElement('form');
    const { target } = formProps;

    form.setAttribute('method', formProps.method || 'POST');
    form.setAttribute('action', formProps.path);

    if (target) {
      form.setAttribute('target', target);
    }

    return form;
  };

  const createSubmitButton = () => {
    const submitButton = document.createElement('input');

    submitButton.setAttribute('type', 'submit');
    submitButton.setAttribute('style', 'display: none;');
    submitButton.setAttribute('value', 'submitButton');

    return submitButton;
  };

  const createInput = (key, value) => {
    const input = document.createElement('input');

    input.setAttribute('type', 'hidden');
    input.setAttribute('name', key);
    input.setAttribute('value', value);

    return input;
  };

  if (formProps && formProps.path) {
    const form = createForm();

    _.each(params, (value, key) => {
      form.appendChild(createInput(key, params[key]));
    });

    form.appendChild(createSubmitButton());
    document.body.appendChild(form);
    form.submit();
    form.parentNode.removeChild(form);
  }
};

export default post;
