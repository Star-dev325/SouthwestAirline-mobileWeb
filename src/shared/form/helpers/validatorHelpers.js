// @flow
import _ from 'lodash';
import validator from 'src/shared/form/formValidators/validator';
import {
  REQUIRED_ERROR,
  DEFAULT_ERROR_TYPE,
  FIELD_ERROR_MESSAGE
} from 'src/shared/form/constants/validationErrorTypes';
import type {
  FormData,
  ValidationRule,
  FormValidationRules,
  FieldValidationRules,
  FormValidationErrors
} from 'src/shared/form/flow-typed/form.types';

function executeRules(rules, ...args) {
  return _.reduce(
    rules,
    (result, { validator: ruleValidator, type = DEFAULT_ERROR_TYPE, msg }) => {
      if (_.isNull(result)) {
        const isValid = ruleValidator && ruleValidator(...args);

        if (_.isPromise(isValid)) {
          const promise = isValid;

          return { promise, type, msg };
        }

        return isValid ? null : { type, msg };
      }

      return result;
    },
    null
  );
}

export function getErrorMsgForField(fieldName: string, fieldValue: *, fieldRules: ValidationRule[]) {
  const { true: requiredRules, false: otherRules } = _.groupBy(fieldRules, (rule) => !!rule.isRequired);
  const isRequiredField = !_.isEmpty(requiredRules);

  if (isRequiredField && !validator.isRequired(fieldValue)) {
    const { msg } = requiredRules[0];

    if (msg) {
      return {
        msg,
        type: FIELD_ERROR_MESSAGE
      };
    } else {
      return {
        type: REQUIRED_ERROR
      };
    }
  }

  if (!isRequiredField && _.isEmpty(fieldValue)) {
    return null;
  }

  return executeRules(otherRules, fieldValue);
}

function hasAnyAsycValidation(errors) {
  return _.some(errors, (error) => _.isPromise(error.promise));
}

function waitAsyncValidationResolvedAndMergeResult(errors) {
  const errorsWithPromise = _.chain(errors)
    .pickBy((error) => _.isPromise(error.promise))
    .entries()
    .value();
  const asyncErrorKeys = _.map(errorsWithPromise, '[0]');
  const asyncErrorContents = _.map(errorsWithPromise, '[1]');

  return Promise.all(_.map(asyncErrorContents, 'promise')).then((asyncErrorValues) => {
    const asyncErrorContentWithValidationResult = _.map(asyncErrorValues, (asyncErrorValue, index: number) => {
      if (!_.isEmpty(asyncErrorValue)) {
        const { type } = asyncErrorContents[index];

        return { type, msg: asyncErrorValue };
      }

      return null;
    });
    const asyncErrors = _.zipObject(asyncErrorKeys, asyncErrorContentWithValidationResult);

    return _.chain(errors).assign(asyncErrors).pickBy(_.negate(_.isEmpty)).value();
  });
}

export function executeValidators(
  formData: FormData,
  formRules: FormValidationRules = {},
  fieldRules: FieldValidationRules = {}
): FormValidationErrors | Promise<FormValidationErrors> {
  const fieldErrors = _.chain(formData)
    .mapValues((fieldValue, fieldName) => getErrorMsgForField(fieldName, fieldValue, fieldRules[fieldName]))
    .pickBy(_.negate(_.isNull))
    .value();

  let errors = fieldErrors;

  if (!_.isEmpty(formRules)) {
    const formErrors = _.chain(formRules)
      .mapValues((rules) => executeRules(rules, formData, fieldErrors))
      .pickBy(_.negate(_.isNull))
      .value();

    errors = _.merge(errors, formErrors);
  }

  if (!hasAnyAsycValidation(errors)) {
    return errors;
  }

  return waitAsyncValidationResolvedAndMergeResult(errors);
}
