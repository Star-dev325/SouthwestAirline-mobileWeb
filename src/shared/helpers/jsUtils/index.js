export * from 'src/shared/helpers/jsUtils/functors';
export * from 'src/shared/helpers/jsUtils/object/index';
export * from 'src/shared/helpers/jsUtils/is';
export * from 'src/shared/helpers/jsUtils/lang/index';
