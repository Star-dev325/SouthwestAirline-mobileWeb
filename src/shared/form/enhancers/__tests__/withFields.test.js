import { fireEvent, render } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import Input from 'src/shared/components/input';
import withFields from 'src/shared/form/enhancers/withFields';
import { noop } from 'src/shared/helpers/jsUtils';

describe('withFields', () => {
  let onChangeStub;

  beforeEach(() => {
    onChangeStub = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should only pass declared form data to child component', () => {
    const { container } = createComponent(['fieldTwo']);

    expect(container).toMatchSnapshot();
  });

  describe('update values', () => {
    it('should update the value to form when calling onChange through child component', () => {
      const { container } = createComponent(['fieldTwo']);

      fireEvent.change(container.querySelector('input'), { target: { value: 'value2' } });

      expect(onChangeStub).toHaveBeenCalledWith('fieldTwo', 'value2');
    });

    it('should not update the value to form when calling onChange with an undeclared field', () => {
      const { container } = createComponent(['fieldOne']);

      fireEvent.change(container.querySelector('input'), { target: { value: 'value2' } });

      expect(onChangeStub).not.toHaveBeenCalled();
    });
  });

  class FormContextProvider extends React.Component {
    static childContextTypes = {
      form: PropTypes.object
    };

    getChildContext() {
      return {
        form: {
          clearError: noop,
          errors: {},
          formData: { fieldOne: 'value1', fieldTwo: 'value2' },
          onChange: this.props.onChange,
          register: noop,
          unregister: noop
        }
      };
    }

    render() {
      return <div>{this.props.children}</div>;
    }
  }

  function createComponent(names = []) {
    const MockedField = (props) =>
      <Input {...props}
        onChange={(e) => { props.onChange('fieldTwo', e.target.value); }}
      />;
    const ConnectedField = withFields(MockedField);

    return render(
      <FormContextProvider onChange={onChangeStub}>
        <ConnectedField names={names} />
      </FormContextProvider>
    );
  }
});
