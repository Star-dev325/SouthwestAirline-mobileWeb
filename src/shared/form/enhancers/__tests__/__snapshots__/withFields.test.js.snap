// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFields should only pass declared form data to child component 1`] = `
<div>
  <div>
    <div
      class="input"
    >
      <input
        class=""
      />
    </div>
  </div>
</div>
`;
